### PR TITLE
Add support for navigation-link descriptions

### LINKED ISSUE #28446: Link description support in Navigation Link block is needed
## What problem does this address?
In the Navigation Link block, there is a field for "Description" in the Link Settings. This was likely added to reflect the description field in the Menus link settings in non-FSE WordPress. Unfortunately, the current Description setting (attribute) does nothing, and there are no filters available in the render function of the block that would allow developers to add support for descriptions. So while the help text indicates that some themes might support descriptions, there does not appear to be a way (currently) for themes to actually support descriptions 😉

image

## What is your proposed solution?
There are a couple solutions that I am happy to explore, but would like feedback from the community on the best path moving forward. 

### Option 1

In core themes, like Twenty Twenty One, link descriptions are not supported, but since the link output is filterable, developers can add support for descriptions as needed. Something as simple as the following would do the trick. 

```
function prefix_nav_descriptions( $item_output, $item, $depth, $args ) {
    if ( ! empty( $item->description ) ) {
        $item_output = str_replace( $args->link_after . '</a>', $args->link_after . '</a><p class="menu-link_description">' . $item->description . '</p>', $item_output );
    }

    return $item_output;
}
add_filter( 'walker_nav_menu_start_el', 'prefix_nav_descriptions', 10, 4 );
```

The simplest approach would be to add similar filters to the navigation-link → index.php file. We could also provide a filter for the submenu icon so it could easily be swapped out. The current implementation is a bit of a headache if you want a different icon.

The problem with this approach is that all the modifications that are made via filters would not be available in the Site/Block Editor. So, while this would be the quickest solution, it is probably not ideal in the long run. 

### Option 2

An alternative approach would be to just make descriptions natively supported in the block. This would take more thought and design, but one implementation could looks something like this. Enabling link descriptions could also a setting in the theme.json file that theme developers could disable as needed. 

image

The link description could also be its own block, but that might start getting too convoluted. 

Let me know what everyone thinks, thanks!

### PR DESCRIPTION
**Pull request by aristath**
## What?
This PR adds support for navigation-link descriptions.
Navigation-link blocks have a `description` field where users can enter a description for their links, and the description of that field is `The description will be displayed in the menu if the current theme supports it.`
However, there is no way for a theme to "support" these.

See #28446

## Why?
Descriptions in menus is a widely supported feature in lots of "classic" themes, but there's currently no way for a block theme to do something similar (unless they rewrite the `render_callback` of the block and override core).
We need a way to allow block themes to use that feature.

## How?
* Add a `<span class="wp-block-navigation-item__description">` after the link's label in the PHP render callback
* Do the same in JS
* Hide the `.wp-block-navigation-item__description` element with CSS. If a theme wants to opt-in to menu descriptions, it is expected they'll do some heavy styling on them, so they can add `display:block` to show them.

### CODE REVIEW

File: packages/block-library/src/navigation-link/style.scss [hunk 1]
```diff
@@ -8,4 +8,10 @@
 		word-break: normal;
 		overflow-wrap: break-word;
 	}
+
+	// Hide the description by default.
+	// If a theme opts-in to show descriptions, they will need to provide styles for them.
+	.wp-block-navigation-item__description {
```
Comment (getdave) [inline packages/block-library/src/navigation-link/style.scss:14]: Why don't we just provide this as an optional attribute via a toggle in the UI (defaulting to `false` "off")? That way you can _choose_ to use it or not. If it's used then it seems only logical that it should be displayed _by default_.

File: packages/block-library/src/navigation-link/edit.js [hunk 2]
```diff
@@ -738,40 +738,47 @@ export default function NavigationLinkEdit( {
...
+										] }
+										onClick={ () => {
+											if ( ! url ) {
+												setIsLinkOpen( true );
+											}
+										} }
+									/>
+									<span className="wp-block-navigation-item__description">
+										{ description || '' }
+									</span>
```
Comment (getdave) [inline packages/block-library/src/navigation-link/edit.js]: Do we want to output DOM if the `description` var is empty?

```suggestion
									{ description && (
									<span className="wp-block-navigation-item__description">
										{ description || '' }
									</span>
									)}
```

**Review by getdave — COMMENTED** [review]
Thanks for working on this. Seems good to actually handle this by default given that the field already exists.

Left some comments and queries.

File: packages/block-library/src/navigation-link/edit.js (hunk 2 above)
Comment (aristath) [reply packages/block-library/src/navigation-link/edit.js]: Good point 👍

File: packages/block-library/src/navigation-link/style.scss (hunk 1 above)
Comment (aristath) [reply packages/block-library/src/navigation-link/style.scss:14]: The problem here is that themes were always opting-in to this feature, by creating their own nav-walkers... Displaying the description usually requires a lot of styling, which usually includes more than just font-sizes & colors. 
Descriptions are _usually_ an opinionated part of a theme's design, so after reading the discussion on https://github.com/WordPress/gutenberg/issues/28446#issuecomment-766307019 it made sense to just hide them and assume that if a theme wants to use description, they're going to add some styling to them - and that can also be used as a way to "opt-in" to the feature.

We _could_ of course add some default styles to descriptions and add a UI switch, but then opting-is is no longer left to the theme, but to users, and that would be a departure from previously established behavior. It's not a bad thing, just different and I don't know if that's something we want to do 🤔

Comment (getdave) [reply packages/block-library/src/navigation-link/style.scss:14]: Let's get some feedback from design folks and see if there are solid opinions on the way to move forward.

Comment (kjellr) [reply packages/block-library/src/navigation-link/style.scss:14]: > We could of course add some default styles to descriptions and add a UI switch, but then opting-is is no longer left to the theme, but to users, and that would be a departure from previously established behavior. It's not a bad thing, just different and I don't know if that's something we want to do ...

I think in general, block themes are already straying from established behavior this way. And for the most part, that's a good thing! FSE is frequently handing more power to users, and preventing features from disappearing on theme switch. So allowing users more control here is definitely in line with those goals. 

To that point, I think the larger question here is: Do we want to support this feature in general? And if so, how would we re-design this to work well alongside other Nav block features? My general impression of this PR is that it seems generally harmless, so I'm not sure whether we need to answer that immediately.

Comment (aristath) [reply packages/block-library/src/navigation-link/style.scss:14]: > To that point, I think the larger question here is: Do we want to support this feature in general?

I think we _have to_ support it... Without it, designs like the one in the screenshot below won't be possible
screenshot of the submenu of the gov.uk site using descriptions for menu items
The screenshot above is one of many examples (this example is from the gov.uk site) where the menu descriptions are essential, both from a design perspective as well as from a UX perspective so that visitors can better comprehend the context of the menu items

Comment (kjellr) [reply packages/block-library/src/navigation-link/style.scss:14]: Yeah, let me rephrase the question: 

Do we want to support this feature *_**in the UI**_* now? And if so, how would we re-design this to work well alongside other Nav block features? 

The alternative of course is basically just what you're doing in the PR already: making this available, but leaving the more intense work for later.

Comment (aristath) [reply packages/block-library/src/navigation-link/style.scss:14]: 😅 apologies, I didn't understand what you meant initially.
Since the field is already there, I believe it would make sense to - as a first step - just make the existing field work.
I think it would be OK to tackle this in stages:
1. Make the existing description field work
2. Integrate the description better in the editor UI
3. Default styling for descriptions

This one is a relatively harmless PR just to get things started, the 1st step towards a "proper" implementation of the descriptions. It doesn't all have to be done in 1 go 😉

Comment (kjellr) [reply packages/block-library/src/navigation-link/style.scss:14]: That makes sense. I think the only risk is that themes would build in custom support for this feature, and then when we get to step 3, our default styles might conflict. But I suppose we can just be careful about specificity when we get to that step. 😅

**Comment by paaljoachim** [comment]
Thank you for creating this PR, Ari!

Testing using Twenty Twenty Two.

Added description.

Backend
Screenshot 2022-04-01 at 11 36 42

Frontend
Screenshot 2022-04-01 at 11 37 41

I expected to see perhaps [Option 2 of the issue](https://github.com/WordPress/gutenberg/issues/28446). Instead I currently do not see any difference.

----

Option 2. Which also includes a toggle to display the description.
In the Appearance -> Menus. Currently one goes to the Screen Options screen to click the checkbox to show the Description field. 

Should we do something like this?

Figma animated file.
https://www.figma.com/proto/JQC8nd7ciOTpprE6ZIbjQN/Untitled?node-id=1%3A2&scaling=min-zoom&page-id=0%3A1&starting-point-node-id=1%3A2

Static mockups.

Screenshot 2022-04-01 at 12 09 36

Screenshot 2022-04-01 at 12 09 47

@jasmussen

**Comment by aristath** [comment]
@paaljoachim This PR adds the HTML for descriptions, both in the editor and the frontend.
So after adding the description, you _should_ have a new `<span class="wp-block-navigation-item__description">` element inside that menu-item.
However, in order to avoid BC the item is hidden by default using `display:none;`. It's up to the theme to properly style descriptions (and therefore add a `display:block` or whatever they want to do in order to have it match their design).

**Review by carolinan — APPROVED** [review]
This works well in my test and I think this should be enough for a first version: It does enable theme developers to show the description which was not possible without this change.

Works best with `align-items: flex-start;` added to `wp-block-navigation__container ` So at some point I would like to see the vertical alignment settings added to the navigation block.


I added the following CSS to style.css in Twenty Twenty-Two, and added flex-start from above:

```
.wp-block-navigation .wp-block-navigation-item__description {
	display: block;
	font-style: italic;
	font-size: .8em;
}
```
navigation block displaying description on menu item

**Comment by paaljoachim** [comment]
It seems my comment would be better added to a new issue in relation to adding the UI for this feature....

**Comment by carolinan** [comment]
After testing this I do feel that the current text area should be enough for the UI.

I mean, when a user has the option to add an alt text for example, they just add their text, we don't ask them to confirm if they want the text or not.

The difference is the "The description will be displayed in the menu if the current theme supports it." Most users will not see any difference with this change.

**Comment by aristath** [comment]
Thank you all for the feedback in this PR!

I'll go ahead and merge it as a 1st step, and we can then iterate further in follow-up PRs to address enhancements. 👍
